File: sdks/typescript/src/crypto.ts
/**
 * Cryptographic operations for O2 Exchange.
 *
 * Provides key generation, wallet creation, and signing operations for
 * both Fuel-native and EVM-compatible wallets. Implements three signing
 * modes used by the O2 Exchange:
 *
 * - **personalSign** — Fuel prefix + SHA-256 (for session creation)
 * - **rawSign** — Plain SHA-256 (for session actions)
 * - **evmPersonalSign** — Ethereum prefix + keccak-256 (for EVM owner sessions)
 *
 * @remarks
 * Uses `@noble/secp256k1` v3 with `prehash:false` since we pre-hash
 * all messages ourselves. The `hmacSha256` configuration is required
 * for synchronous signing.
 *
 * @module
 */

import { hmac } from "@noble/hashes/hmac.js";
import { sha256 } from "@noble/hashes/sha2.js";
import { keccak_256 } from "@noble/hashes/sha3.js";
import * as secp from "@noble/secp256k1";
import { bytesToHex, concat, hexToBytes } from "./encoding.js";

// Configure @noble/secp256k1 v3 for synchronous signing.
// v3 requires manual hash configuration via secp.hashes.
secp.hashes.hmacSha256 = (key: Uint8Array, ...msgs: Uint8Array[]) => {
  const h = hmac.create(sha256, key);
  for (const msg of msgs) h.update(msg);
  return h.digest();
};
secp.hashes.sha256 = (...msgs: Uint8Array[]) => {
  const h = sha256.create();
  for (const msg of msgs) h.update(msg);
  return h.digest();
};

// ── Types ───────────────────────────────────────────────────────────

/**
 * Interface for objects that can sign messages for the O2 Exchange.
 *
 * Both built-in wallets ({@link Wallet}, {@link EvmWallet}) and external
 * signers ({@link ExternalSigner}, {@link ExternalEvmSigner}) satisfy this
 * interface. For custom signing backends (hardware wallets, AWS KMS, HSMs,
 * etc.), implement this interface directly or use the provided external
 * signer classes.
 *
 * @example
 * ```ts
 * // Using the built-in wallet (implements Signer)
 * const wallet = client.generateWallet();
 * const session = await client.createSession(wallet, tradeAccountId, ["fFUEL/fUSDC"]);
 *
 * // Using an external signer
 * const signer = new ExternalSigner("0x1234...abcd", myKmsSignDigest);
 * const session = await client.createSession(signer, tradeAccountId, ["fFUEL/fUSDC"]);
 * ```
 */
export interface Signer {
  /** The Fuel B256 address (0x-prefixed, 64-char hex string). */
  readonly b256Address: string;

  /**
   * Sign a message using the appropriate personal_sign format.
   *
   * For Fuel-native accounts: Fuel personalSign
   * (`\x19Fuel Signed Message:\n` prefix + SHA-256).
   *
   * For EVM accounts: Ethereum personalSign
   * (`\x19Ethereum Signed Message:\n` prefix + keccak256).
   *
   * @param message - The raw message bytes to sign.
   * @returns A 64-byte Fuel compact signature.
   */
  personalSign(message: Uint8Array): Uint8Array;
}

/**
 * Callback type for external signing functions.
 *
 * Receives a 32-byte digest and must return a 64-byte Fuel compact
 * signature (r[32] + s[32] with recovery ID embedded in the MSB of s[0]).
 *
 * Use {@link toFuelCompactSignature} to convert standard `(r, s, recoveryId)`
 * components to the expected format.
 */
export type SignDigestFn = (digest: Uint8Array) => Uint8Array;

/**
 * A Fuel-native secp256k1 wallet.
 *
 * The address is derived as `SHA-256(publicKey[1:65])`, skipping the
 * `0x04` uncompressed prefix byte.
 */
export interface Wallet {
  /** The 32-byte private key. */
  privateKey: Uint8Array;
  /** The 65-byte uncompressed public key (with `0x04` prefix). */
  publicKey: Uint8Array;
  /** The 0x-prefixed, 64-character hex Fuel address. */
  b256Address: string;
}

/**
 * An EVM-compatible secp256k1 wallet.
 *
 * Extends {@link Wallet} with an Ethereum-style address derived as the
 * last 20 bytes of `keccak256(publicKey[1:65])`. The `b256Address` is the
 * EVM address zero-padded to 32 bytes.
 */
export interface EvmWallet extends Wallet {
  /** The 0x-prefixed, 40-character hex Ethereum address. */
  evmAddress: string;
}

// ── Key Generation ──────────────────────────────────────────────────

/**
 * Generate a Fuel-native secp256k1 keypair.
 * Address = SHA-256(publicKey[1:65]) — skip 0x04 prefix.
 */
export function generateWallet(): Wallet {
  const privateKey = secp.utils.randomSecretKey();
  return walletFromPrivateKey(privateKey);
}

/** Load a wallet from a private key (hex string or bytes). */
export function walletFromPrivateKey(privateKeyInput: Uint8Array | string): Wallet {
  const privateKey =
    typeof privateKeyInput === "string" ? hexToBytes(privateKeyInput) : privateKeyInput;
  const publicKey = secp.getPublicKey(privateKey, false); // uncompressed 65 bytes
  const addressBytes = sha256(publicKey.slice(1)); // skip 0x04 prefix
  return {
    privateKey,
    publicKey,
    b256Address: bytesToHex(addressBytes),
  };
}

/**
 * Generate an EVM-compatible secp256k1 keypair.
 * EVM address = last 20 bytes of keccak256(publicKey[1:65]).
 * B256 address = zero-padded EVM address.
 */
export function generateEvmWallet(): EvmWallet {
  const privateKey = secp.utils.randomSecretKey();
  return evmWalletFromPrivateKey(privateKey);
}

/** Load an EVM wallet from a private key. */
export function evmWalletFromPrivateKey(privateKeyInput: Uint8Array | string): EvmWallet {
  const privateKey =
    typeof privateKeyInput === "string" ? hexToBytes(privateKeyInput) : privateKeyInput;
  const publicKey = secp.getPublicKey(privateKey, false);
  const hash = keccak_256(publicKey.slice(1));
  const evmAddressBytes = hash.slice(12); // last 20 bytes

  let evmHex = "0x";
  for (const b of evmAddressBytes) evmHex += b.toString(16).padStart(2, "0");

  // B256: 12 zero bytes + 20 EVM address bytes
  const b256Bytes = new Uint8Array(32);
  b256Bytes.set(evmAddressBytes, 12);

  return {
    privateKey,
    publicKey,
    b256Address: bytesToHex(b256Bytes),
    evmAddress: evmHex,
  };
}

// ── Signing ─────────────────────────────────────────────────────────

/**
 * Sign a 32-byte digest and return a 64-byte Fuel compact signature.
 *
 * Steps:
 * 1. Sign with secp256k1 using prehash:false (digest is already hashed)
 * 2. Low-s normalization is handled by @noble/secp256k1 v3 automatically
 * 3. Embed recovery ID in MSB of byte 32 (first byte of s):
 *    s[0] = (recovery_id << 7) | (s[0] & 0x7F)
 * 4. Return r(32) + s(32) = 64 bytes
 */
export function fuelCompactSign(privateKey: Uint8Array, digest: Uint8Array): Uint8Array {
  // CRITICAL: prehash must be false — the digest is already SHA-256 hashed.
  // Use format: 'recovered' to get 65-byte output: [recovery_id, r(32), s(32)].
  // Low-s normalization is handled automatically by the library.
  const sig65 = secp.sign(digest, privateKey, {
    prehash: false,
    format: "recovered",
  } as Parameters<typeof secp.sign>[2]);
  const recovery = sig65[0]; // recovery ID: 0 or 1
  const r = sig65.slice(1, 33);
  const s = new Uint8Array(sig65.slice(33, 65));

  // Embed recovery ID in MSB of s[0] (first byte of s component)
  s[0] = (recovery << 7) | (s[0] & 0x7f);

  // Return r(32) + s(32) = 64 bytes
  const result = new Uint8Array(64);
  result.set(r, 0);
  result.set(s, 32);
  return result;
}

/**
 * Sign using Fuel's personalSign format (for session creation).
 *
 * prefix = "\x19Fuel Signed Message:\n"
 * digest = sha256(prefix + str(len(message)) + message)
 */
export function personalSign(privateKey: Uint8Array, message: Uint8Array): Uint8Array {
  const prefix = new TextEncoder().encode("\x19Fuel Signed Message:\n");
  const lengthStr = new TextEncoder().encode(String(message.length));
  const fullMessage = concat([prefix, lengthStr, message]);
  const digest = sha256(fullMessage);
  return fuelCompactSign(privateKey, digest);
}

/**
 * Sign using raw SHA-256 hash, no prefix (for session actions).
 *
 * digest = sha256(message)
 */
export function rawSign(privateKey: Uint8Array, message: Uint8Array): Uint8Array {
  const digest = sha256(message);
  return fuelCompactSign(privateKey, digest);
}

/**
 * Sign using Ethereum's personal_sign format (for EVM owner sessions).
 *
 * prefix = "\x19Ethereum Signed Message:\n" + str(len(message))
 * digest = keccak256(prefix + message)
 */
export function evmPersonalSign(privateKey: Uint8Array, message: Uint8Array): Uint8Array {
  const prefix = new TextEncoder().encode(`\x19Ethereum Signed Message:\n${message.length}`);
  const fullMessage = concat([prefix, message]);
  const digest = keccak_256(fullMessage);
  return fuelCompactSign(privateKey, digest);
}

// ── Digest helpers ──────────────────────────────────────────────────

/**
 * Compute the Fuel personalSign digest.
 *
 * Constructs `SHA-256("\x19Fuel Signed Message:\n" + str(len(message)) + message)`
 * and returns the 32-byte digest.
 *
 * Use this when implementing a custom {@link Signer} for Fuel-native accounts.
 */
export function fuelPersonalSignDigest(message: Uint8Array): Uint8Array {
  const prefix = new TextEncoder().encode("\x19Fuel Signed Message:\n");
  const lengthStr = new TextEncoder().encode(String(message.length));
  return sha256(concat([prefix, lengthStr, message]));
}

/**
 * Compute the Ethereum personal_sign digest.
 *
 * Constructs `keccak256("\x19Ethereum Signed Message:\n" + str(len(message)) + message)`
 * and returns the 32-byte digest.
 *
 * Use this when implementing a custom {@link Signer} for EVM accounts.
 */
export function evmPersonalSignDigest(message: Uint8Array): Uint8Array {
  const prefix = new TextEncoder().encode(`\x19Ethereum Signed Message:\n${message.length}`);
  return keccak_256(concat([prefix, message]));
}

/**
 * Convert standard `(r, s, recoveryId)` components to a 64-byte Fuel compact signature.
 *
 * The Fuel compact format stores the recovery ID in the MSB of byte 32
 * (first byte of `s`): `s[0] = (recoveryId << 7) | (s[0] & 0x7F)`.
 *
 * @param r - The 32-byte `r` component of the ECDSA signature.
 * @param s - The 32-byte `s` component (must be low-s normalized).
 * @param recoveryId - The recovery ID (0 or 1).
 * @returns A 64-byte Fuel compact signature.
 *
 * @remarks
 * The `s` component **must be low-s normalized** before calling this function.
 * If `s > secp256k1_order / 2`, negate it (`s = order - s`) and flip the
 * recovery ID (`recoveryId ^= 1`). Most modern signing libraries do this
 * automatically, but check your KMS documentation.
 */
export function toFuelCompactSignature(
  r: Uint8Array,
  s: Uint8Array,
  recoveryId: number,
): Uint8Array {
  const sig = new Uint8Array(64);
  sig.set(r, 0);
  sig.set(s, 32);
  // Embed recovery ID in MSB of s[0]
  sig[32] = (recoveryId << 7) | (sig[32] & 0x7f);
  return sig;
}

// ── External Signers ────────────────────────────────────────────────

/**
 * A Fuel-native signer backed by an external signing function.
 *
 * Use this for hardware wallets, AWS KMS, or other secure enclaves
 * that manage private keys externally. The SDK handles Fuel-specific
 * message framing (prefix + SHA-256 hashing); your callback only needs
 * to sign a raw 32-byte digest.
 *
 * @example
 * ```ts
 * import { ExternalSigner, toFuelCompactSignature } from "@o2exchange/sdk";
 *
 * const signer = new ExternalSigner("0x1234...abcd", (digest) => {
 *   const { r, s, recoveryId } = myKms.sign(digest);
 *   return toFuelCompactSignature(r, s, recoveryId);
 * });
 *
 * const session = await client.createSession(signer, tradeAccountId, ["fFUEL/fUSDC"]);
 * ```
 */
export class ExternalSigner implements Signer {
  /** The Fuel B256 address (0x-prefixed hex string). */
  readonly b256Address: string;
  private readonly signDigest: SignDigestFn;

  /**
   * @param b256Address - The Fuel B256 address (0x-prefixed, 64-char hex).
   * @param signDigest - Callback that signs a 32-byte digest and returns
   *   a 64-byte Fuel compact signature.
   */
  constructor(b256Address: string, signDigest: SignDigestFn) {
    this.b256Address = b256Address;
    this.signDigest = signDigest;
  }

  /**
   * Sign using Fuel's personalSign format, delegating to the external signer.
   *
   * Computes `SHA-256("\x19Fuel Signed Message:\n" + len + message)` and
   * passes the 32-byte digest to the `signDigest` callback.
   */
  personalSign(message: Uint8Array): Uint8Array {
    const digest = fuelPersonalSignDigest(message);
    return this.signDigest(digest);
  }
}

/**
 * An EVM signer backed by an external signing function.
 *
 * Same as {@link ExternalSigner} but uses Ethereum personal_sign message
 * framing (`\x19Ethereum Signed Message:\n` prefix + keccak256 hashing).
 *
 * @example
 * ```ts
 * import { ExternalEvmSigner, toFuelCompactSignature } from "@o2exchange/sdk";
 *
 * const signer = new ExternalEvmSigner(
 *   "0x000000000000000000000000abcd...1234", // b256 (zero-padded)
 *   "0xabcd...1234",                          // EVM address
 *   (digest) => {
 *     const { r, s, recoveryId } = myKms.sign(digest);
 *     return toFuelCompactSignature(r, s, recoveryId);
 *   },
 * );
 * ```
 */
export class ExternalEvmSigner implements Signer {
  /** The Fuel B256 address (zero-padded EVM address). */
  readonly b256Address: string;
  /** The EVM address (0x-prefixed, 40-char hex). */
  readonly evmAddress: string;
  private readonly signDigest: SignDigestFn;

  /**
   * @param b256Address - The B256 address (EVM address zero-padded to 32 bytes).
   * @param evmAddress - The EVM address (0x-prefixed, 40-char hex).
   * @param signDigest - Callback that signs a 32-byte digest and returns
   *   a 64-byte Fuel compact signature.
   */
  constructor(b256Address: string, evmAddress: string, signDigest: SignDigestFn) {
    this.b256Address = b256Address;
    this.evmAddress = evmAddress;
    this.signDigest = signDigest;
  }

  /**
   * Sign using Ethereum's personal_sign format, delegating to the external signer.
   *
   * Computes `keccak256("\x19Ethereum Signed Message:\n" + len + message)` and
   * passes the 32-byte digest to the `signDigest` callback.
   */
  personalSign(message: Uint8Array): Uint8Array {
    const digest = evmPersonalSignDigest(message);
    return this.signDigest(digest);
  }
}

export { hexToBytes, bytesToHex };

File: sdks/rust/src/crypto.rs
use secp256k1::ecdsa::RecoverableSignature;
/// Cryptographic operations for O2 Exchange: key generation, signing, and address derivation.
///
/// Implements:
/// - Fuel-native key generation (SHA-256 address derivation)
/// - EVM key generation (keccak256 address derivation)
/// - personalSign (Fuel prefix + SHA-256)
/// - rawSign (plain SHA-256)
/// - evm_personal_sign (Ethereum prefix + keccak256)
/// - fuel_compact_sign with low-s normalization and recovery ID in MSB of byte 32
use secp256k1::{Message, PublicKey, Secp256k1, SecretKey};
use sha2::{Digest as Sha256Digest, Sha256};
use sha3::Keccak256;

use crate::errors::O2Error;

/// Half of the secp256k1 group order, used for low-s normalization.
const SECP256K1_ORDER_HALF: [u8; 32] = [
    0x7F, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
    0x5D, 0x57, 0x6E, 0x73, 0x57, 0xA4, 0x50, 0x1D, 0xDF, 0xE9, 0x2F, 0x46, 0x68, 0x1B, 0x20, 0xA0,
];

/// Full secp256k1 group order.
const SECP256K1_ORDER: [u8; 32] = [
    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE,
    0xBA, 0xAE, 0xDC, 0xE6, 0xAF, 0x48, 0xA0, 0x3B, 0xBF, 0xD2, 0x5E, 0x8C, 0xD0, 0x36, 0x41, 0x41,
];

/// A Fuel-native wallet with SHA-256 derived B256 address.
#[derive(Debug, Clone)]
pub struct Wallet {
    pub private_key: [u8; 32],
    pub public_key: [u8; 65],
    pub b256_address: [u8; 32],
}

/// An EVM-compatible wallet with keccak256 derived address, zero-padded to B256.
#[derive(Debug, Clone)]
pub struct EvmWallet {
    pub private_key: [u8; 32],
    pub public_key: [u8; 65],
    pub evm_address: [u8; 20],
    pub b256_address: [u8; 32],
}

/// Generate a Fuel-native secp256k1 keypair.
/// Address = SHA-256(uncompressed_pubkey[1..65])
pub fn generate_keypair() -> Result<Wallet, O2Error> {
    let secp = Secp256k1::new();
    let mut rng = rand::thread_rng();
    let (secret_key, public_key) = secp.generate_keypair(&mut rng);

    let pubkey_bytes = public_key.serialize_uncompressed();
    let address = Sha256::digest(&pubkey_bytes[1..65]);

    Ok(Wallet {
        private_key: secret_key.secret_bytes(),
        public_key: pubkey_bytes,
        b256_address: address.into(),
    })
}

/// Generate an EVM-compatible keypair.
/// EVM address = last 20 bytes of keccak256(uncompressed_pubkey[1..65])
/// B256 address = 12 zero bytes + 20 EVM address bytes
pub fn generate_evm_keypair() -> Result<EvmWallet, O2Error> {
    let secp = Secp256k1::new();
    let mut rng = rand::thread_rng();
    let (secret_key, public_key) = secp.generate_keypair(&mut rng);

    let pubkey_bytes = public_key.serialize_uncompressed();
    let keccak_hash = Keccak256::digest(&pubkey_bytes[1..65]);

    let mut evm_address = [0u8; 20];
    evm_address.copy_from_slice(&keccak_hash[12..32]);

    let mut b256_address = [0u8; 32];
    b256_address[12..32].copy_from_slice(&evm_address);

    Ok(EvmWallet {
        private_key: secret_key.secret_bytes(),
        public_key: pubkey_bytes,
        evm_address,
        b256_address,
    })
}

/// Load a Fuel-native wallet from a private key.
pub fn load_wallet(private_key: &[u8; 32]) -> Result<Wallet, O2Error> {
    let secp = Secp256k1::new();
    let secret_key = SecretKey::from_slice(private_key)
        .map_err(|e| O2Error::CryptoError(format!("Invalid private key: {e}")))?;
    let public_key = PublicKey::from_secret_key(&secp, &secret_key);
    let pubkey_bytes = public_key.serialize_uncompressed();
    let address = Sha256::digest(&pubkey_bytes[1..65]);

    Ok(Wallet {
        private_key: *private_key,
        public_key: pubkey_bytes,
        b256_address: address.into(),
    })
}

/// Load an EVM wallet from a private key.
pub fn load_evm_wallet(private_key: &[u8; 32]) -> Result<EvmWallet, O2Error> {
    let secp = Secp256k1::new();
    let secret_key = SecretKey::from_slice(private_key)
        .map_err(|e| O2Error::CryptoError(format!("Invalid private key: {e}")))?;
    let public_key = PublicKey::from_secret_key(&secp, &secret_key);
    let pubkey_bytes = public_key.serialize_uncompressed();
    let keccak_hash = Keccak256::digest(&pubkey_bytes[1..65]);

    let mut evm_address = [0u8; 20];
    evm_address.copy_from_slice(&keccak_hash[12..32]);

    let mut b256_address = [0u8; 32];
    b256_address[12..32].copy_from_slice(&evm_address);

    Ok(EvmWallet {
        private_key: *private_key,
        public_key: pubkey_bytes,
        evm_address,
        b256_address,
    })
}

/// Compare two 32-byte big-endian numbers: returns true if a > b.
fn gt_be(a: &[u8; 32], b: &[u8; 32]) -> bool {
    for i in 0..32 {
        if a[i] > b[i] {
            return true;
        }
        if a[i] < b[i] {
            return false;
        }
    }
    false
}

/// Negate a 32-byte big-endian number modulo the secp256k1 order.
/// result = ORDER - value
fn negate_s(s: &[u8; 32]) -> [u8; 32] {
    let mut result = [0u8; 32];
    let mut borrow: u16 = 0;
    for i in (0..32).rev() {
        let diff = SECP256K1_ORDER[i] as u16 - s[i] as u16 - borrow;
        result[i] = diff as u8;
        borrow = if diff > 255 { 1 } else { 0 };
    }
    result
}

/// Sign a 32-byte digest and return a 64-byte Fuel compact signature.
///
/// The recovery ID is embedded in the MSB of byte 32 (first byte of s).
/// Low-s normalization is applied: if s > order/2, negate s and flip recovery_id.
pub fn fuel_compact_sign(private_key: &[u8; 32], digest: &[u8; 32]) -> Result<[u8; 64], O2Error> {
    let secp = Secp256k1::new();
    let secret_key = SecretKey::from_slice(private_key)
        .map_err(|e| O2Error::CryptoError(format!("Invalid private key: {e}")))?;
    let message = Message::from_digest(*digest);

    let recoverable_sig: RecoverableSignature = secp.sign_ecdsa_recoverable(&message, &secret_key);
    let (rec_id, compact) = recoverable_sig.serialize_compact();
    let mut recovery_id = rec_id.to_i32() as u8;

    let mut r = [0u8; 32];
    let mut s = [0u8; 32];
    r.copy_from_slice(&compact[0..32]);
    s.copy_from_slice(&compact[32..64]);

    // Low-s normalization
    if gt_be(&s, &SECP256K1_ORDER_HALF) {
        s = negate_s(&s);
        recovery_id ^= 1;
    }

    // Embed recovery ID in MSB of s[0]
    s[0] = (recovery_id << 7) | (s[0] & 0x7F);

    let mut result = [0u8; 64];
    result[0..32].copy_from_slice(&r);
    result[32..64].copy_from_slice(&s);
    Ok(result)
}

/// Sign using Fuel's personalSign format (for session creation).
/// prefix = b"\x19Fuel Signed Message:\n" + str(len(message)) + message
/// digest = sha256(prefix)
pub fn personal_sign(private_key: &[u8; 32], message: &[u8]) -> Result<[u8; 64], O2Error> {
    let prefix = b"\x19Fuel Signed Message:\n";
    let length_str = message.len().to_string();

    let mut hasher = Sha256::new();
    hasher.update(prefix);
    hasher.update(length_str.as_bytes());
    hasher.update(message);
    let digest: [u8; 32] = hasher.finalize().into();

    fuel_compact_sign(private_key, &digest)
}

/// Sign using raw SHA-256 hash, no prefix (for session actions).
/// digest = sha256(message)
pub fn raw_sign(private_key: &[u8; 32], message: &[u8]) -> Result<[u8; 64], O2Error> {
    let digest: [u8; 32] = Sha256::digest(message).into();
    fuel_compact_sign(private_key, &digest)
}

/// Sign using Ethereum's personal_sign format (for EVM owner session creation).
/// prefix = "\x19Ethereum Signed Message:\n" + str(len(message))
/// digest = keccak256(prefix_bytes + message)
pub fn evm_personal_sign(private_key: &[u8; 32], message: &[u8]) -> Result<[u8; 64], O2Error> {
    let prefix = format!("\x19Ethereum Signed Message:\n{}", message.len());

    let mut hasher = Keccak256::new();
    hasher.update(prefix.as_bytes());
    hasher.update(message);
    let digest: [u8; 32] = hasher.finalize().into();

    fuel_compact_sign(private_key, &digest)
}

/// Trait for wallets that can sign messages for O2 Exchange operations.
///
/// Implemented for both [`Wallet`] (Fuel-native, SHA-256) and [`EvmWallet`] (keccak256).
pub trait SignableWallet {
    /// The B256 address used as the owner identity.
    fn b256_address(&self) -> &[u8; 32];
    /// Sign a message using the wallet's personal_sign scheme.
    ///
    /// - Fuel wallets use `\x19Fuel Signed Message:\n` prefix + SHA-256.
    /// - EVM wallets use `\x19Ethereum Signed Message:\n` prefix + keccak256.
    fn personal_sign(&self, message: &[u8]) -> Result<[u8; 64], O2Error>;
}

impl SignableWallet for Wallet {
    fn b256_address(&self) -> &[u8; 32] {
        &self.b256_address
    }
    fn personal_sign(&self, message: &[u8]) -> Result<[u8; 64], O2Error> {
        personal_sign(&self.private_key, message)
    }
}

impl SignableWallet for EvmWallet {
    fn b256_address(&self) -> &[u8; 32] {
        &self.b256_address
    }
    fn personal_sign(&self, message: &[u8]) -> Result<[u8; 64], O2Error> {
        evm_personal_sign(&self.private_key, message)
    }
}

/// Derive a Fuel B256 address from a public key (65 bytes, 0x04 prefix).
pub fn address_from_pubkey(public_key: &[u8; 65]) -> [u8; 32] {
    Sha256::digest(&public_key[1..65]).into()
}

/// Derive an EVM address from a public key (65 bytes, 0x04 prefix).
pub fn evm_address_from_pubkey(public_key: &[u8; 65]) -> [u8; 20] {
    let hash = Keccak256::digest(&public_key[1..65]);
    let mut addr = [0u8; 20];
    addr.copy_from_slice(&hash[12..32]);
    addr
}

/// Format a 32-byte array as a "0x"-prefixed hex string.
pub fn to_hex_string(bytes: &[u8]) -> String {
    format!("0x{}", hex::encode(bytes))
}

/// Parse a "0x"-prefixed hex string into a 32-byte array.
pub fn parse_hex_32(s: &str) -> Result<[u8; 32], O2Error> {
    let s = s.strip_prefix("0x").unwrap_or(s);
    let bytes = hex::decode(s).map_err(|e| O2Error::CryptoError(format!("Invalid hex: {e}")))?;
    if bytes.len() != 32 {
        return Err(O2Error::CryptoError(format!(
            "Expected 32 bytes, got {}",
            bytes.len()
        )));
    }
    let mut result = [0u8; 32];
    result.copy_from_slice(&bytes);
    Ok(result)
}
